perf: add pagination in logs api

deepanshupal09-datazip opened the pull request:
# Description

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR implements cursor based pagination in the logs API in both frontend and BFF.

- FE: Implement logs virtualisation and bi-directional infinite scrolling using react-virtuoso
- FE: Implement new logs API in the frontend
- BFF: Implement cursor and limit based pagination in the logs API.

This ensures smooth UX for user and reduces load on the client browser by only loading logs that are currently visible by the user.

## Type of change

<!--
Please delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- [x] Tested the BFF API using various cursor and limit and the edge cases
- [x] Tested the FE, ensured the logs appear in correct order.

# Screenshots or Recordings
<!-- Attach related screenshots or recordings here -->

## Related PR's (If Any):

vishalm0509 commented on server/utils/utils.go:
```diff
@@ -245,52 +452,120 @@ func ReadLogs(mainLogDir string) ([]map[string]interface{}, error) {
 	logPath := filepath.Join(logDir, "olake.log")
-	logContent, err := os.ReadFile(logPath)
+
+	f, err := os.Open(logPath)
```
can we use a meaningful name ?

vishalm0509 commented on server/internal/handlers/job.go:
```diff
@@ -373,9 +374,13 @@ func (h *Handler) GetTaskLogs() {
-	logger.Debugf("Get task logs initiated job_id[%d] file_path[%s]", id, req.FilePath)
+	cursor, _ := h.GetInt64("cursor", constants.DefaultLogsCursor)
+	limit, _ := h.GetInt("limit", constants.DefaultLogsLimit)
+	direction := h.GetString("direction", constants.DefaultLogsDirection)
```
Do we need to handle error ? let's say ?cursor=abc. 

just a suggestion, check if we need it

```go
   if err != nil && h.GetString("cursor") != "" {
        utils.ErrorResponse(&h.Controller, http.StatusBadRequest, "invalid cursor parameter: must be a number", err)
        return
    }
```

vishalm0509 commented on ui/src/store/taskStore.ts:
```diff
@@ -43,13 +66,34 @@ export const createTaskSlice: StateCreator<TaskSlice> = set => ({
+			taskLogsOlderCursor: LOGS_CONFIG.DEFAULT_CURSOR,
+			taskLogsNewerCursor: LOGS_CONFIG.DEFAULT_CURSOR,
+			taskLogsHasMoreOlder: true,
+			taskLogsHasMoreNewer: true,
```
Newer = true ?

vishalm0509 commented on ui/src/store/taskStore.ts:
```diff
@@ -61,4 +105,119 @@ export const createTaskSlice: StateCreator<TaskSlice> = set => ({
+				filePath,
+				taskLogsNewerCursor,
+				LOGS_CONFIG.SUBSEQUENT_BATCH_SIZE,
+				"newer",
```
can we make it `"old" and "new"` ? wdyt

vishalm0509 commented on ui/src/modules/jobs/pages/JobLogs.tsx:
```diff
@@ -28,82 +35,183 @@ const JobLogs: React.FC = () => {
+	}, [taskLogs, searchText, showOnlyErrors])
+
+	useEffect(() => {
+		if (!jobs.length) fetchJobs()
```
Should we do this in a separate useEffect() ?

deepanshupal09-datazip replied:
done

vishalm0509 commented on ui/src/modules/jobs/pages/JobLogs.tsx:
```diff
@@ -185,95 +289,81 @@ const JobLogs: React.FC = () => {
+									if (!log) return null
+									const item = log as any
+									const hasTimeField = Boolean(item.time)
+									return (
```
Can we create this as a component ?

this `JobLogss` Component is pretty big

vishalm0509 commented on server/utils/utils.go:
```diff
@@ -245,52 +452,120 @@ func ReadLogs(mainLogDir string) ([]map[string]interface{}, error) {
+		for _, line := range lines {
+			var logEntry LogEntry
+
+			if err := json.Unmarshal([]byte(line), &logEntry); err != nil {
```
The `lines` is already validated by unmarshalling in the `isValidLogLine`, so this is redundant. Please check once

deepanshupal09-datazip commented on server/utils/utils.go:
```diff
@@ -225,8 +230,210 @@ type LogEntry struct {
+// Filters out empty lines, invalid JSON, and debug-level logs DURING reading.
+// startOffset is treated as exclusive - we read lines that END BEFORE startOffset.
+// Returns: valid lines (oldest->newest), newOffset (byte position before first returned line), hasMore, error.
+func ReadLinesBackward(f *os.File, startOffset int64, limit int) ([]string, int64, bool, error) {
```
I also thought of this... but this is decreasing time on cost of memory as we are maintaining a separate data structure for storing line with it's offset. Time complexity is gonna be same anyways but your approach will be slightly faster and will use slightly more memory, should we go with this?

deepanshupal09-datazip replied:
fixed this

deepanshupal09-datazip commented on ui/src/modules/jobs/pages/JobLogs.tsx:
```diff
@@ -28,82 +35,183 @@ const JobLogs: React.FC = () => {
-
 	const { Search } = Input
 
+	const START_INDEX = 1000000 // High starting index to allow space for prepending older log entries without disrupting Virtuoso's list virtualization
```
done

deepanshupal09-datazip commented on ui/src/modules/jobs/pages/JobLogs.tsx:
```diff
@@ -28,82 +35,183 @@ const JobLogs: React.FC = () => {
 			}
+			return matchesSearch
+		})
+	}, [taskLogs, searchText, showOnlyErrors])
```
fixed

deepanshupal09-datazip commented on server/utils/utils.go:
```diff
@@ -225,8 +229,205 @@ type LogEntry struct {
+			currentOffset += int64(len(lineBytes))
+
+			// Remove trailing newline and check if valid
+			line := strings.TrimRight(string(lineBytes), "\r\n")
```
Not needed as `json.Unmarshal` handles trailing newlines, but we can keep it for consistency as in `ReadLineBackwards` we do `data[lastNL+1:]`, which excludes the newline.

What do you suggest?

deepanshupal09-datazip commented on ui/src/modules/jobs/pages/JobLogs.tsx:
```diff
@@ -28,82 +37,194 @@ const JobLogs: React.FC = () => {
+		) {
+			setTimeout(() => {
+				virtuosoRef.current?.scrollToIndex({
+					index: filteredLogs.length - 1,
```
length won't be zero as we check above for this...

```
	// Initial Scroll to Bottom
	useEffect(() => {
		if (
			!isLoadingTaskLogs &&
			filteredLogs?.length > 0 &&
			!hasPerformedInitialScroll.current
		) {
```

deepanshupal09-datazip commented on server/utils/utils.go:
```diff
@@ -225,8 +229,205 @@ type LogEntry struct {
+	fileSize := fi.Size()
+
+	// Clamp startOffset to [0, fileSize]
+	startOffset = max(startOffset, 0)
```
done... updated the comment

deepanshupal09-datazip commented on server/utils/utils.go:
```diff
@@ -225,8 +229,205 @@ type LogEntry struct {
+func ReadLogs(mainLogDir string, cursor int64, limit int, direction string) (*dto.TaskLogsResponse, error) {
 	// Check if mainLogDir exists
 	if _, err := os.Stat(mainLogDir); os.IsNotExist(err) {
 		return nil, fmt.Errorf("logs directory not found: %s", mainLogDir)
```
included `err` in `fmt.Error` now

deepanshupal09-datazip commented on server/utils/utils.go:
```diff
@@ -245,52 +446,120 @@ func ReadLogs(mainLogDir string) ([]map[string]interface{}, error) {
+				case string:
+					messageStr = v
+				default:
+					msgBytes, _ := json.Marshal(v)
```
done

deepanshupal09-datazip commented on ui/src/modules/jobs/pages/JobLogs.tsx:
```diff
@@ -28,82 +37,194 @@ const JobLogs: React.FC = () => {
+			filteredLogs?.length > 0 &&
+			!hasPerformedInitialScroll.current
+		) {
+			setTimeout(() => {
```
done

deepanshupal09-datazip commented on ui/src/modules/jobs/pages/JobLogs.tsx:
```diff
@@ -185,95 +302,42 @@ const JobLogs: React.FC = () => {
+								startReached={handleStartReached}
+								endReached={handleEndReached}
+								firstItemIndex={firstItemIndex}
+								overscan={1000}
```
done

vishalm0509 commented:
> this is intended behaviour... we have to fire it 2-3 times because backend sometimes still returns old tasks, and not newer ones, when you come to this page by clicking "Sync Now".

Can we not fire it always, and retry based on if the desired job is present or not. 

```go
for (let i = 0; i < 5; i++) {
  const tasks = await fetchTasks();
  if (tasks.find(t => t.type === 'sync' && t.createdAt > startTime)) {
    break;
  }
  await sleep(300 * i); // exponential wait
}
```

Something like this, but not exponential wait. I think this will stop the extra fires if available at the first response. 

wdyt ?

deepanshupal09-datazip commented:
> > this is intended behaviour... we have to fire it 2-3 times because backend sometimes still returns old tasks, and not newer ones, when you come to this page by clicking "Sync Now".
> 
> Can we not fire it always, and retry based on if the desired job is present or not.
> 
> ```go
> for (let i = 0; i < 5; i++) {
>   const tasks = await fetchTasks();
>   if (tasks.find(t => t.type === 'sync' && t.createdAt > startTime)) {
>     break;
>   }
>   await sleep(300 * i); // exponential wait
> }
> ```
> 
> Something like this, but not exponential wait. I think this will stop the extra fires if available at the first response.
> 
> wdyt ?

fixed this

deepanshupal09-datazip commented:
> A request is being triggered for small files.
> 
>  Screen.Recording.2025-12-11.at.9.18.40.AM.mov

thanks for pointing this out, it was due to an edge case where we have read all the valid logs, but there are some other logs (debug logs) that we are not including... so it was returning hasMore true but there are no valid logs above that cursor value

fixed this issue, now we return hasMore false and cursor as 0 if no valid logs are above that cursor, to prevent that extra API call

.... same thing is also fixed for ReadLinesForward too.

vishalm0509 commented on ui/src/modules/jobs/pages/Jobs.tsx:
```diff
@@ -38,7 +38,12 @@ const Jobs: React.FC = () => {
+				},
+			}) // navigate to job history so that user can see the tasks running
 			await jobService.syncJob(id)
 			await fetchJobs()
```
We are fetching `fetchJobs` in the History page also. Do we need this here ?

```ts
	if (!jobs.length) {
			fetchJobs()
		}
```

deepanshupal09-datazip replied:
> We are fetching `fetchJobs` in the History page also. Do we need this here ?
> 
> ```ts
> 	if (!jobs.length) {
> 			fetchJobs()
> 		}
> ```

because we need to show to job name there

deepanshupal09-datazip replied:
> Why are we navigating and then starting the sync ?

Idk man `¯\_(ツ)_/¯` , fixed this

vishalm0509 commented on server/utils/utils.go:
```diff
@@ -225,72 +229,345 @@ type LogEntry struct {
+				startPos: 0, // First line starts at position 0
+			})
+		}
+	}
```
Can we keep this inside this check ?

```ts
if offset == 0 {
            break
        }
```

deepanshupal09-datazip commented on ui/src/modules/jobs/pages/JobHistory.tsx:
```diff
@@ -51,22 +62,43 @@ const JobHistory: React.FC = () => {
 					retryCountRef.current++
 					timeoutId = setTimeout(fetchWithRetry, THROTTLE_DELAY)
 				} else {
 					setIsDelayingCall(false)
```
done

vishalm0509 reviewed (approved):
LGTM
